pkg/controller/cronjob/cronjob_controller2.go

```diff
@@ -0,0 +1,580 @@
+			return
+		}
+		jm.enqueueController(cronJob)
+		return
+	}
+}
+
+// updateJob figures out what CronJob(s) manage a Job when the Job
+// is updated and wake them up. If the anything of the Job have changed, we need to
+// awaken both the old and new CronJob. old and cur must be *batchv1.Job
+// types.
+func (jm *Controller2) updateJob(old, cur interface{}) {
+	curJob := cur.(*batch.Job)
+	oldJob := old.(*batch.Job)
+	if curJob.ResourceVersion == oldJob.ResourceVersion {
+		// Periodic resync will send update events for all known jobs.
+		// Two different versions of the same jobs will always have different RVs.
+		return
+	}
+
+	curControllerRef := metav1.GetControllerOf(curJob)
+	oldControllerRef := metav1.GetControllerOf(oldJob)
+	controllerRefChanged := !reflect.DeepEqual(curControllerRef, oldControllerRef)
+	if controllerRefChanged && oldControllerRef != nil {
+		// The ControllerRef was changed. Sync the old controller, if any.
+		if cronJob := jm.resolveControllerRef(oldJob.Namespace, oldControllerRef); cronJob != nil {
+			jm.enqueueController(cronJob)
+		}
+	}
+
+	// If it has a ControllerRef, that's all that matters.
+	if curControllerRef != nil {
+		cronJob := jm.resolveControllerRef(curJob.Namespace, curControllerRef)
+		if cronJob == nil {
+			return
+		}
+		jm.enqueueController(cronJob)
+		return
+	}
+}
+
+func (jm *Controller2) deleteJob(obj interface{}) {
+	job, ok := obj.(*batchv1.Job)
+
+	// When a delete is dropped, the relist will notice a job in the store not
+	// in the list, leading to the insertion of a tombstone object which contains
+	// the deleted key/value. Note that this value might be stale.
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("couldn't get object from tombstone %#v", obj))
+			return
+		}
+		job, ok = tombstone.Obj.(*batchv1.Job)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("tombstone contained object that is not a ReplicaSet %#v", obj))
+			return
+		}
+	}
+
+	controllerRef := metav1.GetControllerOf(job)
+	if controllerRef == nil {
+		// No controller should care about orphans being deleted.
+		return
+	}
+	cronJob := jm.resolveControllerRef(job.Namespace, controllerRef)
+	if cronJob == nil {
+		return
+	}
+	jm.enqueueController(cronJob)
+}
+
+func (jm *Controller2) enqueueController(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %+v: %v", obj, err))
+		return
+	}
+
+	jm.queue.Add(key)
+}
+
+func (jm *Controller2) enqueueControllerWithTime(obj interface{}, t time.Duration) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %+v: %v", obj, err))
+		return
+	}
+
+	jm.queue.AddAfter(key, t)
+}
+
+func (jm *Controller2) updateCronJob(old interface{}, curr interface{}) {
+	//  the big cases that needs to be handled
+	//  1. change in CJ.spec.schedule
```

Comment:
Only this matters here, since this affects your queue. The other `.spec` elements apply to the next execution, that's how the current controller works and that's what users will expect, that shouldn't change.

Reply:
Also the comment is different than what you do below :wink:

Reply:
transferred the comments to the inline doc and updated to reflect what the function does accurately